### ISSUE TITLE
Restic backup — OOM-killed on raspberry pi after backing up another computer to same repo

### ISSUE THREAD

**Issue by hmage** [restic/restic#1988]
On a raspberry pi 3 model B, which has 1GB of memory, restic stopped fitting into RAM after a 3TB backup was done on another machine to same repository. I suspect indexes have grown big enough to not fit into memory.

## Output of `restic version`
```
hmage@phmd:~$ restic version
restic 0.9.2 compiled with go1.10.3 on linux/arm
```

## How did you run restic exactly?
```bash
export AWS_ACCESS_KEY_ID=secret
export AWS_SECRET_ACCESS_KEY=secret
export RESTIC_PASSWORD=secret
export RESTIC_REPOSITORY=s3:https://s3.wasabisys.com/restic-hmage
source ./restic-excludes.sh
restic backup --exclude-file <(printf "%s\n" "${EXCLUDES[@]}") /`
```

Contents of `restic-excludes.sh`:
```bash
EXCLUDES=(
/dev
/proc
/run
/sys
)
EXCLUDES+=(
$'Icon\r'
$HOME/.bundle
$HOME/.cache
$HOME/.cargo
$HOME/.ccache*
$HOME/.config/chromium
$HOME/.cpan
$HOME/.dropbox
$HOME/.local/share/akonadi
$HOME/.npm
$HOME/Library/Application\ Support/Google/Chrome
$HOME/Library/Application\ Support/Telegram\ Desktop
$HOME/Library/Arq/Cache.noindex
$HOME/norm.*
**/var/cache/apt
**/var/cache/man
**/var/lib/apt/lists
**/var/lib/mlocate
.DS_Store
.DocumentRevisions-V100
.Spotlight-V100
.Trashes
.bzvol
.cache
.dropbox.cache
.fseventsd
/Volumes/Time\ Machine
/media/psf
/private/var/vm/
/srv/piwik/tmp
/srv/www/data/cache
/tmp
/usr/lib/debug
/var/lib/lxcfs
/var/swap
/var/tmp
Cache
Caches
)
```

## What backend/server/service did you use to store the repository?
Wasabi (S3 protocol)

## Expected behavior
Restic should not run out of memory no matter how big the indexes are — they should be streamed from disk/repo rather than loaded completely into RAM since it's not infinite.

Actual behavior
---------------
Restic allocates tons of memory depending of index sizes — before backing up 3TB of data on my mac, restic on Pi had no problems backing up, after successful backup, restic gets killed by kernel's oom-killer:

```
hmage@phmd:~$ dmesg|fgrep restic
[426681.565821] [15683]  1000 15683     1393      295       8       0        0             0 restic-backup.s
[426681.565827] [15709]  1000 15709   268547   174432     353       0        0             0 restic
[426681.565897] Out of memory: Kill process 15709 (restic) score 664 or sacrifice child
[426681.565959] Killed process 15709 (restic) total-vm:1074188kB, anon-rss:697728kB, file-rss:0kB, shmem-rss:0kB
[426681.766777] oom_reaper: reaped process 15709 (restic), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
[511937.088184] [17227]  1000 17227     1393      295       6       0        0             0 restic-backup.s
[511937.088190] [17255]  1000 17255   267651   176005     357       0        0             0 restic
[511937.088205] Out of memory: Kill process 17255 (restic) score 670 or sacrifice child
[511937.088266] Killed process 17255 (restic) total-vm:1070604kB, anon-rss:704020kB, file-rss:0kB, shmem-rss:0kB
[511937.324251] oom_reaper: reaped process 17255 (restic), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
[598842.337688] [25297]  1000 25297     1393      295       5       0        0             0 restic-backup.s
[598842.337695] [25324]  1000 25324   201281   129430     264       0        0             0 restic
[598842.337735] Out of memory: Kill process 25324 (restic) score 493 or sacrifice child
[598842.337793] Killed process 25324 (restic) total-vm:805124kB, anon-rss:517720kB, file-rss:0kB, shmem-rss:0kB
[598842.529990] oom_reaper: reaped process 25324 (restic), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
[642004.263182] [25392]  1000 25392     1409      314       5       0        0             0 restic-backup.s
[642004.263188] [25412]  1000 25412   201122   123536     252       0        0             0 restic
[642004.263252] Out of memory: Kill process 25412 (restic) score 470 or sacrifice child
[642004.263305] Killed process 25412 (restic) total-vm:804488kB, anon-rss:494144kB, file-rss:0kB, shmem-rss:0kB
[642004.409938] oom_reaper: reaped process 25412 (restic), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```

Steps to reproduce the behavior
-------------------------------

1. backup on Pi into empty repo anywhere — will succeed
2. backup from another machine 3TB of data with 2 million files, which takes about several attempts and about 20 hours.
3. backup on Pi again into same repo — will OOM

Do you have any idea what may have caused this?
-----------------------------------------------
Restic tries to load all indexes into RAM instead of mmaping them.

Do you have an idea how to solve the issue?
-------------------------------------------
mmap the indexes from disk cache

Did restic help you or made you happy in any way?
-------------------------------------------------
So far best backup solution, and only backup solution that allows deduplicate backups from across multiple machines into single repo.

**Comment by fd0** [comment]
Yep, scaling is hard. One of the issues we currently have is that the index (which blobs are stored where) is loaded into RAM. Our plans to improve are:
 * Make the information loaded into memory smaller
 * Maybe create a temporary file on disc for holding other data

**Comment by joshdreagan** [comment]
I had the same issue. I initially ran a very large backup locally on the backup machine itself so that it would complete faster. I then tried to run the same backup (which would be incremental) from an odroid with about 2GB of RAM and ran out of memory in about 30 seconds. As a workaround, I just created a swapfile. Slows things down a bit, but got me past the memory issue.

```
# run as root
fallocate -l 8G /swapfile
chmod 600 /swapfile
mkswap /swapfile
swapon /swapfile
```

**Comment by fd0** [comment]
> Also, thank you so much for restic! It is the best backup program I've ever used.

Thank you, such comments make my day :)

> and running the part that holds the indexes in RAM on a big server that has tons of disk space and RAM available.

That'd introduce a huge amount of complexity, splitting restic into two parts. But thanks for the idea!

**Comment by fd0** [comment]
@pzmarzly I'm sorry if this sounds blunt, but I'm going to be direct here :)

While I appreciate the suggestion (I'm familiar with what `mmap` does and what it's used for), it feels to me that you don't have the full picture, so a solution is applied before understanding what's going on exactly. In this case, `mmap` will not improve the situation at all, in addition to the problems with using `mmap` in Go you listed yourself.

restic works roughly as follows:
 * It reads index files from the disc. We've limited the size of the index files to limit the amount of memory that is used.
 * It verifies the integrity, checks the signature, and decrypts the content (also into memory)
 * The JSON data contained in the file is parsed and its content is inserted into an in-memory data structure which is used for lookups during runtime

So `mmap` would only help for the very first step of this process. At a minimum, decryption and parsing will still need to happen in memory.

All steps described above use memory only for a short amount of time, then the Go garbage collector claims and frees it again. The culprit here is the data structure in memory, which keeps the data in a way that lets it grow very large. That needs to be addressed, everything else will just slightly improve the symptoms.

In addition, it feels to me that something with the way we use our data structures makes it very hard for the runtime's garbage collector to actually free memory, I haven't had the time to dig into this yet.

**Comment by xacrimon** [comment]
@fd0 would you be interested in me collecting some gc stats while running? It would show if the gc has problems freeing memory. I'd gladly run some when I come home

**Comment by fd0** [comment]
You can do that, please let me know if you find anything interesting. I already know that there's plenty of memory to be garbage collected, but I don't understand why the runtime still seems to use more. So, if you like to dig in, please do so! I even have a sample repo you can use, which consists of a few hundred snapshots of Linux kernel source code (so plenty of small files), the repo size is around 50GiB. Please let me know if you'd like to get it.

**Comment by xacrimon** [comment]
@fd0 I've done extensive testing by now and it seems like the core of the problem is the Gc not being able to efficiently detect unused memory

**Comment by fd0** [comment]
What do you mean by "freed"? It is normal that memory the GC detected as "not in use any more" is not instantly given back to the OS. Sometimes Linux even prefers to have the memory allocated to the process, depending on the memory pressure.

FYI: this is likely something different than the original issue described by @hmage: they could not load the index fully into memory on the small machine. Which is probably caused by inefficient structures (memory-wise).

**Comment by xacrimon** [comment]
Freed as in go can reuse this memory

**Comment by joshtriplett** [comment]
I'm running into the same issue, attempting to back up a server with 184873 files, 7.593 GiB total, with 722MB of RAM (some of which is in use).

I'd be happy to run tests to help debug this.

**Comment by joshtriplett** [comment]
As a workaround, `export GOGC=20` before running restic substantially reduces memory usage.

**Comment by tcurdt** [comment]
@fd0 memory usage is a main blocker for me to adopt `restic`

Could you elaborate on this a bit?

> So mmap would only help for the very first step of this process. At a minimum, decryption and parsing will still need to happen in memory.

How much needs to be in memory? Just a block I assume?
So what breaks scalability is really only the index?

**Comment by MarkMielke** [comment]
I see the same problem as this when backing up a number of Git repositories. The "restic backup" process takes several GiB. We can increase the VM size that it runs on, but this seems like a scaling issue.

The GOGC=20 option helped. In this case, it seems to basically mean that Restic needs up to 120% of the RAM needed to hold the indices, vs GOGC=100 (default) which means that Restic needs up to 200% of the RAM needed to hold the indices. So, GOGC=20 result in a huge improvement in limiting the effect, but doesn't solve the underlying problem.

**Comment by aawsome** [comment]
To tackle the problem I'd like to summarize the technical details and also start a discussion about possible solutions.

- Basically the main problem with most commands is the index hold by `internal/repository` that holds the information of all index files in memory. The memory consumption thus grows (approximately) linear with the number of blobs in the repository.
- Number of blobs in the repository can grow hugely for several reasons, e.g. high number of (small) files, generally high backup volume (much data) or backup from many computers to one repository.
- The commands `rebuild-index` and `prune` additionally use an in-memory index built in `internal/index` which makes the memory situation even worse (besides performance problems with these commands).

So I see several approaches, we can take into consideration:

## rebuild-index and prune

About the `rebuild-index` and `prune` a re-engineering should be done using the index from 'internal/repository' and thus removing the double in-memory index creation. A starting point could be PR #2513 which just uses 'repository/index' and already adds a prune-like capability. A capability to compact packs (maybe 'compack-packs') is missing but should be also able to implement just using 'repository/index'. Also a repair functionality for missing index data can IMO be done without the 'internal/index' functionality.

## Reduce memory usage per blob

About the issue with high-memory-usage of many blobs, I see several posibilites (most of which have already been discussed):

1. With PR #2338 there is already a proposal optimizing for the case that blobs are usually only stored in one pack.
2. I would like to generally discuss if we can omit the information about more than one pack per blob and just use the first pack.
3. Going more aggressive could mean to just hold the full information for tree blobs in memory. For data I would suggest to only save the information that a blob is stored. If the information "where" is needed for a specific data blob reread all index files until the blob is found. This approach would not affect most commands but slow down the commands which actually need to read data blobs like 'restore'.
4. Maybe not storing the complete index at all and instead reading all index files for every index operation is also an alternative. Maybe add a bounded in-memory cache for blob already read. (no idea about performance impacts of this proposal...) This is similar to storage-based index, see below.

## Use storage-based index

This approach can be seen as an alternative to reducing the memory used per blob. Here instead of writing the complete index to an in-memory data structure the index is saved in a database on disc.
I think a simple key/value storage would make the job.
There are some implementations out, e.g.:

- [bbolt](https://github.com/etcd-io/bbolt)
- [moss](https://github.com/couchbase/moss)
- [diskv](https://github.com/peterbourgon/diskv)
- other candidates?

## Backup from many computers to one repository

In order to allow *many* computers to backup to one repository without increasing I would like to propose the following idea:

- Each computer uses its own list of index files
- That means the index is not build by reading all index files but instead of the 
- The list of index files has to be stored somewhere. Possibilities are: extra path in the repository layout to keep track of computers or save the list in the snapshot file.
- This will most likely "kill" deduplication between computers
- To gain again deduplication, index rebuilds and pack reorganization has to be done using all index files, e.g. on a separate computer with large memory.

## Separate index database from computer backing up

This idea implies to separate restic in two parts which can be run on separate hardware. The "storage backend" part would be in duty to manage index files and actually read from and write to the storage. This can be on a computer with large memory. The "frontend" part would just use an API to get the information if blobs are existing or to read blobs and write blobs, respectively. This is similar to how borg is working.
As @fd0 already pointed out, this would be a complete redesign of the restic architecture.

## Discussion

Are there any ideas I am missing?
What do you think about the proposals?
Which direction is the most promising one to start with?

Looking forward to an open and fruitful discussion!

**Comment by irasnyd** [comment]
Hey @aawsome, thanks for the excellent write up! I'm only a user of restic, but I'll throw in my own two cents. This memory usage issue is the primary reason I'm unable to use restic, so it's important to me. (I'm also a software developer, but not a Golang developer.)

Users expect to use tons of disk space these days. 10TB and larger hard drives are readily available, and are <$200 where I live. Multi-TB hard drives are standard items even on very cheap low-spec workstations. RAM is much more expensive, and users don't have nearly as much of it. The same cheap low-spec workstations usually only have 8GB RAM. The amount of RAM in workstations isn't growing as rapidly as the amount of disk space available. Users will only create more data in the future, not less.

From my point of view, this issue is a risk to the restic project's longer term future and survival.

My use-case is that I want to back up both user workstations (home directories only) as well as larger file servers. The user workstations home directories are usually fairly small (50 to 100 GB) but might contains **lots** of small files. I have some users that have tens or hundreds of millions of small <1KByte files. Users do weird stuff sometimes. The larger file servers are up to about 20 TB, but I'm looking towards 100TB in the next year. Mostly larger files (50 to 100 MB each), but there are still a lot of files to fill up that much space.

I think a bounded in-memory cache is a great idea. I also like the idea of a storage-based index.

The ideal storage based index for my use case would be one which grows proportionally to the amount of files backed up *from this specific computer only*. If the disk space needed for the index grows proportionally to the amount of files backed up in total across all computers, I'll hit the same issue we have now, but with disk space instead of RAM.

I intentionally put small (<250GB) disks in user workstations to force them to use the managed file servers as much as possible.

I like your idea about how to split the repository for use by many computers. For my own use case, I would be willing to give up deduplication between computers. I would find it acceptable to do a repack or index rebuild on a large computer once in a while. I'd also be perfectly happy with no support for deduplication between computers at all with this type of multi-computer repository.

I think your idea about separating the index from computer backup through an API is a good long term plan. If this were implemented, I'd find it easy to justify a larger machine to run the API with more RAM, but especially a larger faster SSD to hold the on-disk storage-based index. (A consumer grade 1TB SSD is ~$100 where I live. 1TB RAM is much much much more expensive than that.) This would entirely remove both disk and RAM usage from clients completely.

As a fellow software developer, I understand the amount of work involved, and how big of a task and challenge that would be.

I hope this explains my use cases and the situations where I've hit this issue myself. Please feel free to ask any follow up questions. I know how to compile Golang programs, and can perform real world experiments for you. I have access to a large restic backup repository stored in AWS S3 storage. It is currently ~56TB. Please let me know if there is anything I can help with to move this issue along!

I hope this reply comes off as positive. I think everyone on the restic team has done absolutely terrific work. It is the best backup program I've used in my entire career.

**Comment by shtratos** [comment]
I've recently started using Restic. Once I read the design document I was stunned by the beauty of its simplicity. I really want my restore/backup tool to be as simple as possible - it makes it much easier to trust it and reason about it. Thank you very much for creating such a beautiful thing.

The following is strictly my opinion, please correct me if I'm wrong.


## Scaling issues

I really appreciate that Restic is small and works out of the box for simple use cases.
However, it's evident that it doesn't scale for large number of blobs.
I'm definitely interested in making Restic use fixed amount of memory for all operations.

Thank you @aawsome for the great summary of the current scaling issues. 

Indices are the biggest one, but there are others I saw:
 - a directory tree blob is composed and encrypted in memory regardless of number of files in it #2446
 - REST server does not support paging and may perform unbounded amount of work per request #2421

I think we should keep them in mind while discussing scaling solutions but fix them one-by-one.


## Scaling and scoping strategy

IMO we really need guidance from the core maintainers here:
 - making Restic scale will inevitably increase its complexity.
   Some users require scaling while others would prefer simplicity.
   This feels like 2 different products to me.
 - I got the impression that self-tuning is preferred to allowing users to configure all the things.
 - I saw in many threads that Restic team wants to limit scope of what Restic does
   e.g. no backup scheduling, no additional backends, no GUI
   I fully support that decision, Unix way FTW!
   It would be great if there is clear guidance on what belongs in Restic core and what should be separate.
   Is scaling to arbitrary number of blobs and any repo size in scope?


## Data collection

I understand that it's hard to make strategic decisions in absense of usage data.
It'd be good if we can conduct a census of existing Restic repos.
Maybe add a command that sends anonymous stats about backups, restores and repository to the team?

Information I'd like to see:
 - repo size on disk
 - number of blobs: tree & data
 - number of different machines backing up to the repository (is that information even stored anywhere?)
 - history of how many new files were backed up and how many new blobs were created
 - history of memory consumption during backup and restore
 - max number of files in directory
 - max number of blobs in a file
 - max file size
 - histogram of file and blob sizes in a repo

Obviously collection of this stats should be entirely optional, but it will help to make informed decisions.

## Storage-based Index

Storage-based index (whether local or remote) seems like a good long term solution to me.

As I looked at the code, `Index` interface is very narrow (which is good!):

```go
type Index interface {
	Has(ID, BlobType) bool
	Lookup(ID, BlobType) ([]PackedBlob, bool)
	Count(BlobType) uint

	Each(ctx context.Context) <-chan PackedBlob
}
```

I didn't find any usages of `Count` outside tests, so we need only point-lookup and iteration over all elements.
Storage index needs to be encrypted to provide the same guarantees as existing indices.

As a proof of concept of scalable backup command with minimal changes to current code:
 - continue to fetch index files from the backend
 - provide `Index` implementation that reads index files and stores index entries in the local database instead of memory
 - use the database for lookups - let it handle the caching and what's kept in memory or on disk.
 - for simplicity delete the database after restic command finishes
 - configure the database to use fixed amount of memory

Hopefully users can use this POC and provide feedback on its performance.

As another potential storage solution, I suggest to consider [Badger](https://github.com/dgraph-io/badger).
It's written in go, embeddable, allows to configure memory usage and supports encryption at rest.

Long term I think it'd be great if we can expose `Index` as a gRPC or REST API and allow plugging in different implementations. Then the complexity and maintenance costs can live outside Restic core.


**TL;DR** scaling requires trading off simplicity, scaling strategy is needed, let's build a storage-based index prototype

I'm really excited about this project and happy to discuss this further.

**Comment by ViRb3** [comment]
I wonder if Go's new GC might have a positive impact on restic: https://go.dev/blog/greenteagc

**Comment by MichaelEischer** [comment]
I wouldn't expect much of a benefit. The GC is already able to completely skip the index during garbage collection.
